### PR TITLE
Create estimate-cvap.py

### PR DESCRIPTION
**Pull request by ccthegreat313**
This code is largely unchanged from when Anthony originally created it. Other than method-izing it, the following changes are reflected in this PR: 
* percentage_cap paramter: This is the cap for the CVAP to VAP ratio, i.e. when given CVAP for a certain group for 2019 and VAP for the same group for 2020, the ratio could potentially be > 1. Originally, for any value > 1, the ratio was reset to 1; now the user can define at what ratio they want to reset to 1. 
* zero_fill parameter: When there is 0 CVAP in a geographic region, the ratio will be reset to zero_fill.
* If percentage_cap and zero_fill are not passed as parameters when this function is called, the user will see a value error. 
* If there are nan values when calculating the CVAP to VAP ratio, the nans get filled with the average CVAP to VAP ratio in the county. This assumes that the identifier for the dataframe containing CVAP will be a census geometry identifier which has county fips at indices [2:5] within it. For more explanation on this, go here: https://www.census.gov/programs-surveys/geography/guidance/geo-identifiers.html. I don't know if this limits the code; if someone can pull CVAP using a different method than what's in the data sub-folder this may create a problem. 

I would love some feedback on lines 52-61 where I do the neighboring average calculation, just because this seems messy to me and if there's a better way to do it, I'd love to know! Also any other feedback is greatly appreciated (of course), but I anticipate editing those lines, specifically, before this is merged in.

### CODE REVIEW

**Review by gabeschoenbach — COMMENTED** [review]
Great to have this in! Here are my thoughts so far:
 * should we just make `percentage_cap` and `zero_fill` required parameters, rather than default to `None`?
 * I wonder if we should rename `cvap_source_identifier` to something like `cvap_geoid` and make it default to "TRACT10". That way there is flexibility to change it, but we're signalling that it's really meant to be used to match 2020 blocks to 2010 tracts. If we do this, we could also spell that out more explicitly in the docstring for `base`, and say something like "must contain a GEOID column that matches the `cvap_geoid`, usually "TRACT10".

I think everything from lines 39 and down look good — I don't totally know if the county stuff at lines 52-61 could be cleaned up or not, would be interested what other people think. But more broadly, I wonder what the pros/cons are of putting the `cvap()` and/or `acs5()` calls _inside_ of this `estimate_cvap()` function itself. So then the user wouldn't pass a CVAP source, but rather the base data (with *VAP counts on 2020 blocks) and a list of subpopulations for which they want 2020 CVAP estimates for. Something like 
```python
estimate_cvap(df, ["BCVAP", "HCVAP", "WCVAP"])
```
and then the function could pull down the relevant columns from the ACS/CVAP functions and do the calculations necessary. One thing that is hard about this method is that we'd have to parse the user's list of subpopulations and figure out what the relevant ACS columns are, rather than the way done here, which hardcodes in the columns. I'm not super tied to doing it this way, but its what I was picturing at first and I wonder what people think.

**Comment by gabeschoenbach** [comment]
OK, just added the modification that makes the user input a `cvap_groups` list that would look like this, for example: `[("CVAP19", "VAP19", "VAP20"), ("BCVAP19", "BVAP19", "BVAP20"), ("NHBCVAP19", "BVAP19", "BVAP20")]`.

File: evaltools/data/estimatecvap.py [hunk 1]
```diff
@@ -0,0 +1,258 @@
+
+import pandas as pd
+import geopandas as gpd
+import numpy as np
+from pandas import DataFrame
+from .acs import acs5, cvap
+from ..geometry import unitmap
+
+
+def fetchgeometries(state, geometry) -> gpd.GeoDataFrame:
```
Comment (apizzimenti) [inline evaltools/data/estimatecvap.py:10]: Changed the positional args order for consistency.

File: evaltools/data/estimatecvap.py [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+    if geometry == "block group": geometry = "bg"
+
+    # Construct the Census URL.
+    fips = state.fips
+    head = "https://www2.census.gov/geo/pvs/tiger2010st/"
+    tail = f"{fips}_{clocator}/{fips}/tl_2010_{fips}_{geometry}10.zip"
+    url = head + tail
+
+    # Download, extract, and return the geometries from the URL.
+    return gpd.read_file(url)
```
Comment (apizzimenti) [inline evaltools/data/estimatecvap.py:39]: `geopandas` can read straight from URLs!

File: evaltools/data/estimatecvap.py [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+    
+    # Get the right name and rename the 2010 geometry index this way.
+    geometry10id = "TRACT10" if geometry == "tract" else "BLOCKGROUP10"
+    geometry10 = geometry10.rename({"GEOID10": geometry10id}, axis=1)
+    
+    # Create a unit mapping from the provided base units to those retrieved from
+    # the Census. If the `base` passed has been sliced or could possibly be a
+    # copy, *this will throw a SettingWithCopy warning*.
+    mapping = unitmap((base, baseindex), (geometry10, geometry10id))
+    base[geometry10id] = base[baseindex].map(mapping)
```
Comment (apizzimenti) [inline evaltools/data/estimatecvap.py:67]: Uses `.map()` to get 2010 GEOIDs from the mapping.

File: evaltools/data/estimatecvap.py [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+        geometry10 = "tract"
+
+    # Grab ACS and CVAP special-tab data, and make sure our triples are correct
+    cvap_geoid = "TRACT10" if geometry10 == "tract" else "BLOCKGROUP10"
+    acs_source = acs5(state, geometry10)
+    cvap_source = cvap(state, geometry10)
+
+    # Validate the columns passed, issuing user warnings when it's inadvisable
+    # to estimate CVAP given the passed columns.
+    for (cvap10, vap10, vap20) in groups:
```
Comment (apizzimenti) [inline evaltools/data/estimatecvap.py:129]: Just adding some returns so formatting is consistent across machines and doesn't get wrapped accidentally!

File: evaltools/data/estimatecvap.py [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+    acs_source = acs_source[[cvap_geoid] + non_overlaps]
+    source = cvap_source.merge(acs_source, on=cvap_geoid)
+
+    # Get the right columns from the base geometry, and map the base geometries
+    # to the units with CVAP data. Apparently dropping bad columns by using slicing
+    # incudes a SettingWithCopy warning, so we're just dropping using .drop()
+    # instead.
+    correct = ["geometry"] + [col for col in list(base) if any(sub in col for sub in ["POP", "VAP", "GEOID"])]
+    bads = list(set(base) - set(correct))
+    pared = base.drop(bads, axis=1)
```
Comment (apizzimenti) [inline evaltools/data/estimatecvap.py:174]: Slicing induces a pandas `SettingWithCopyWarning`, so we `.drop()` bad columns instead.

File: evaltools/data/estimatecvap.py [hunk 6]
```diff
@@ -0,0 +1,258 @@
...
+    # Construct the Census URL.
+    fips = state.fips
+    head = "https://www2.census.gov/geo/pvs/tiger2010st/"
+    tail = f"{fips}_{clocator}/{fips}/tl_2010_{fips}_{geometry}10.zip"
+    url = head + tail
+
+    # Download, extract, and return the geometries from the URL.
+    return gpd.read_file(url)
+
+def mapbase(base, state, geometry, baseindex="GEOID20"):
```
Comment (apizzimenti) [inline evaltools/data/estimatecvap.py:41]: Slightly modified call signature for consistency!

File: evaltools/data/estimatecvap.py [hunk 7]
```diff
@@ -0,0 +1,258 @@
...
+    for (cvap10, vap10, _) in groups: 
+       source[cvap10 + "%"] = source[cvap10]/source[vap10]
+
+    # Fill in values according to the following rules:
+    # 
+    #   1.  if there are 0 *CVAP reported and 0 *VAP reported, we set the weight to
+    #       the average *CVAP/*VAP ratio within the county;
+    #   2.  if there are 0 *CVAP reported and *VAP > 0, we set the weight to zero_fill;
+    #   3.  if *CVAP > 0 but *VAP = 0 or *CVAP/*VAP > percentage_cap, we set the weight to 1.
+    statewide = {
```
Comment (apizzimenti) [inline evaltools/data/estimatecvap.py:187]: This may end up being useless – still worth it to keep in?

File: evaltools/data/estimatecvap.py [hunk 8]
```diff
@@ -0,0 +1,258 @@
...
+    countyaverages = { pct: {} for pct in cvappcts }
+
+    for county in counties:
+        chunk = source[source["_county"] == county]
+        
+        for cvap19, vap19, _ in groups:
+            cvap19total = chunk[cvap19].sum()
+            vap19total = chunk[vap19].sum()
+            countyaverages[cvap19 + "%"][county] = cvap19total/vap19total
+
```
Comment (apizzimenti) [inline evaltools/data/estimatecvap.py:208]: Here we compute the county averages for each CVAP column only once, and can then apply them based on the county identifier of each `base` unit.

File: evaltools/data/estimatecvap.py [hunk 9]
```diff
@@ -0,0 +1,258 @@
...
+    assert all(
+        np.all(source[p + "%"] <= ceiling)
+        for (p, _, _) in groups
+    )
+    
+    # Assert we don't have any NaNs or zeros.
+    assert not all(
+        source[p + "%"].isnull().values.any() and np.all(source[p + "%"] > 0)
+        for (p, _, __) in groups
+    )
```
Comment (apizzimenti) [inline evaltools/data/estimatecvap.py:233]: Adding an extra `assert` to ensure our `NaN` and zeroes are properly filled.

File: evaltools/data/estimatecvap.py [hunk 10]
```diff
@@ -0,0 +1,258 @@
...
+
+    # Set indices and create a mapping from IDs to weights.
+    source = source.set_index(cvap_geoid)    
+    source = source[cvappcts]
+    weights = source.to_dict(orient="index")
+
+    # Group by the CVAP GEOID.
+    groupedtogeometry = list(pared.groupby(cvap_geoid))
```
Comment (apizzimenti) [inline evaltools/data/estimatecvap.py:241]: Slight renaming because I got confused.

File: evaltools/geometry/unitmap.py [hunk 11]
```diff
@@ -36,6 +36,7 @@ def unitmap(source, target) -> dict:
     # Set a progress bar; filter out all warnings; create the mapping.
     maup.progress.enabled = True
     warnings.simplefilter("ignore", UserWarning)
+    warnings.simplefilter("ignore", FutureWarning)
```
Comment (apizzimenti) [inline evaltools/geometry/unitmap.py:39]: Ignore `FutureWarning`s from somewhere in `maup`.

File: tests/test_data.py [hunk 12]
```diff
@@ -4,16 +4,39 @@
...
-    remap
+    remap, estimatecvap
 )
 import us
+import geopandas as gpd
+import math
 
 root = Path(os.getcwd()) / Path("tests/test-resources/")
 
+def test_estimate_cvap():
```
Comment (apizzimenti) [inline tests/test_data.py:18]: Added a test where I grabbed two 2010 tracts geographically distant from each other, matched them to 2020 block groups, and hand-calculated the correct values. The tests check whether the results in the dataframe match those expected results!

**Review by apizzimenti — APPROVED** [review]
Some small revisions (mainly just re-ordering parameters in call signatures, renaming variables, adding tests, etc.) but this looks good to go.

**Comment by apizzimenti** [comment]
OH I forgot to mention: `fetchgeometries()` is an incredibly useful method, so I think we should generalize that to get 2020 geometries too!

**Pull request merged**